src/btree/bt_cursor.c

```diff
@@ -1253,11 +1264,37 @@ __btcur_update(WT_CURSOR_BTREE *cbt, WT_ITEM *value, u_int modify_type)
 	WT_ERR(__cursor_localvalue(cursor));
 	__cursor_state_save(cursor, &state);
 
-retry:	WT_ERR(__cursor_func_init(cbt, true));
+	/*
+	 * If our caller configures for a local search and we have a page
+	 * pinned, do that search.
+	 */
+	if (F_ISSET(cursor, WT_CURSTD_UPDATE_LOCAL)
+	    && __cursor_page_pinned(cbt)) {
+		__wt_txn_cursor_op(session);
+		WT_ERR(__wt_txn_autocommit_check(session));
+		WT_ERR(btree->type == BTREE_ROW ?
+		    __cursor_row_search(cbt, true, cbt->ref, &leaf_found) :
+		    __cursor_col_search(cbt, cbt->ref, &leaf_found));
+		/*
+		 * Only use the pinned page search results if search returns an
+		 * exact match or a slot other than the page's boundary slots,
+		 * if that's not the case, a neighbor page might offer a better
+		 * match. This test is simplistic as we're ignoring append
+		 * lists (there may be no page slots or we might be
+		 * legitimately positioned after the last page slot). Ignore
+		 * those cases, it makes things too complicated.
+		 */
+	if (leaf_found && (cbt->compare == 0
+	    || (cbt->slot != 0 && cbt->slot != cbt->ref->page->entries - 1)))
+		goto update_local;
+    }
 
+retry:	WT_ERR(__cursor_func_init(cbt, true));
```

Comment:
Maybe just put the `__cursor_func_init` call on a newline? The extra indentation is a bit misleading because it implies scope.

Reply:
resolved.